bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpThingHandler.java

```diff
@@ -0,0 +1,251 @@
+                    } catch (IllegalArgumentException | IllegalStateException e) {
+                        logger.warn("Failed processing REFRESH command for channel {}: {}", channelUID, e.getMessage());
+                    }
+                }
+            }
+        } else {
+            try {
+                itemValueConverter.send(command);
+            } catch (IllegalArgumentException e) {
+                logger.warn("Failed to convert command '{}' to channel '{}' for sending", command, channelUID);
+            } catch (IllegalStateException e) {
+                logger.debug("Writing to read-only channel {} not permitted", channelUID);
+            }
+        }
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(HttpThingConfig.class);
+
+        if (config.baseURL.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "Parameter baseURL must not be empty!");
+            return;
+        }
+
+        if (!config.username.isEmpty()) {
+            try {
+                authentication = new BasicAuthentication.BasicResult(new URI(config.baseURL), "username", "password");
+            } catch (URISyntaxException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                        "failed to create authentication");
+            }
+        }
+        thing.getChannels().forEach(this::createChannel);
+
+        updateStatus(ThingStatus.ONLINE);
+    }
+
+    @Override
+    public void dispose() {
+        // stop update tasks
+        urlHandlers.values().forEach(RefreshingUrlCache::stop);
+
+        // clear lists
+        urlHandlers.clear();
+        channels.clear();
+        channelUrls.clear();
+
+        super.dispose();
+    }
+
+    /**
+     * create all necessary information to handle every channel
+     *
+     * @param channel a thing channel
+     */
+    private void createChannel(Channel channel) {
+        ChannelUID channelUID = channel.getUID();
+
+        HttpChannelConfig channelConfig = channel.getConfiguration().as(HttpChannelConfig.class);
+
+        String stateUrl = concatenateUrlParts(config.baseURL, channelConfig.stateExtension);
+        String commandUrl = channelConfig.commandExtension == null ? stateUrl
+                : concatenateUrlParts(config.baseURL, channelConfig.commandExtension);
+
+        ItemValueConverter itemValueConverter;
+
+        switch (channel.getAcceptedItemType()) {
+            case "Color":
+                itemValueConverter = new ColorItemConverter(state -> updateState(channelUID, state),
+                        command -> sendCommand(commandUrl, command),
+                        valueTransformationProvider.getValueTransformation(channelConfig.stateTransformation),
+                        valueTransformationProvider.getValueTransformation(channelConfig.commandTransformation),
+                        channelConfig);
```

Comment:
I hate code duplication, but even I don't know if what I'm suggesting is worth it...

```suggestion
                itemValueConverter = createConverter(ColorItemConverter::new, channelUID, channelConfig);
```

```
public ItemValueConverter createConverter(ItemConverterFactory factory, ChannelUID channelUID,
HttpChannelConfig channelConfig){
        String stateUrl = concatenateUrlParts(config.baseURL, channelConfig.stateExtension);
        String commandUrl = channelConfig.commandExtension == null ? stateUrl
                : concatenateUrlParts(config.baseURL, channelConfig.commandExtension);

     return factory.create(state -> updateState(channelUID, state),
                        command -> sendCommand(commandUrl, command),
                        valueTransformationProvider.getValueTransformation(channelConfig.stateTransformation),
                        valueTransformationProvider.getValueTransformation(channelConfig.commandTransformation),
                        channelConfig);
}


@FunctionalInterface
private interface ItemConverterFactory{
    ItemValueConverter create(Consumer<State> updateState, @Nullable Consumer<String> sendCommand,
            ValueTransformation stateTransformations, ValueTransformation commandTransformations,
            HttpChannelConfig channelConfig);
}
```